### ISSUE TITLE
Handling shutdown

### ISSUE THREAD

**Issue by lhstrh** [icyphy/lingua-franca#62]
> \martin{To stop periodic timers (and maybe kill other pending events in the queue)
> we need a request-for-shutdown function.}
> \marten{I'm not convinced that we do. Can't the base reactor class simply implement
> a reaction that does this upon being triggered by a shutdown action?}
> \martin{How do you stop a periodic timer? Who decides that a shutdown happens?
> We should have an example application to explore this.}
> \edward{I suggest moving this discussion to the bug tracking system.}

In the CyPhy paper we assume that the container of a reactor schedules shutdown when appropriate. If this does not satisfactorily address the question, we could explore alternatives.

**Comment by schoeberl** [comment]
> On 21 Nov, 2019, at 20:38, Marten Lohstroh <notifications@github.com> wrote:
> 
> \martin{To stop periodic timers (and maybe kill other pending events in the queue)
> we need a request-for-shutdown function.}
> \marten{I'm not convinced that we do. Can't the base reactor class simply implement
> a reaction that does this upon being triggered by a shutdown action?}
> \martin{How do you stop a periodic timer? Who decides that a shutdown happens?
> We should have an example application to explore this.}
> \edward{I suggest moving this discussion to the bug tracking system.}
> 
> In the CyPhy paper we assume that the container of a reactor schedules shutdown when appropriate. If this does not satisfactorily address the question, we could explore alternatives.
> 
> 

What does it mean “when appropriate”? The LF wiki talks about a stop() method that can be called.

Martin

**Comment by lhstrh** [comment]
Execution can be requested to halt by calling a runtime function, as explained in the wiki. _When_ it is called, the shutdown sequence unfolds conform the CyPhy paper, by letting the top-level reactor schedule a shutdown action for all reactors contained by it, which then also schedule shutdown for the reactors that they contain, etc. This goes on until the entire reactor hierarchy has reacted to a `shutdown` action.

**Comment by cmnrd** [comment]
I am a bit confused by this statement. What does 'which **then** also schedule shutdown' mean? Are all the shutdown reactions logically simultaneous or is there a microstep delay between the shutdown reactions of different hierarchy levels?

**Comment by lhstrh** [comment]
All actions, including `shutdown` are always scheduled with a microstep
delay. The only exception is `startup`, which is scheduled without a
microstep delay.

**Comment by cmnrd** [comment]
Yes, of course. But what I meant is something else. As you explained above, the shutdown action of all reactors contained by top is scheduled with a microstep delay. What I don't understand from your statment is, how the subsequent scheduling of contained reactors work. Does the scheduling of shutdown for contained reactors introduce an additional microstep delay so that shutdown actions in different hierarchical levels are scheduled with different tags or are all shutdown actions scheduled with the same tag? Or in other words: Does `stop()` schedule all shutdown actions at once (at the same logical time) or does each reactor schedule shutdown of contained reactors in its reaction to shutdown?

**Comment by lhstrh** [comment]
Just commenting on this because I realized that I never properly answered your question. The answer below is more complete than I could have given you when you asked the question. The truth is that there were inconsistencies between the various runtime implementations, and the formalization was incomplete. We've revisited our design slightly, mostly because we've gained a better understanding of how we want mutations to work. Some work is still needed to update our run time environments (I've already done this for TypeScript, @edwardalee  is currently working on the C runtime, I believe).

Neither `startup` nor `shutdown` can be scheduled. `schedule` is only used to create events in the future.

However, reactors can be created or deleted -- at the beginning or end of the program, or during execution, as part of a mutation. Importantly, all reactions have a dependency on the last mutation of their container. This ensures that structural changes made by mutations cannot interfere with reactions in the container. This rule also provides a natural reason for disallowing mutations to depend on the output of reactions of contained reactors (it would result in a zero-delay feedback).

The creation of a reactor is instantaneous. It causes the triggering of any of its reactions sensitive to `startup` or any timer that has an offset of zero. The deletion of a reactor is also instantaneous, in the sense it gets labeled as "defunct" and any of its reactions triggered by `shutdown` executed logically instantaneously. When a reactor gets deleted, all of its contained reactors are automatically (and instantaneously) deleted as well. The actual removal (cleanup) of defunct reactors happens _after_ they have had a chance to react to `shutdown`, at the end of the logical tick at which they were deleted. This is an extra task that has to be performed at the end of each tick.

Finally, ordinary reactions do have access to `stop()`, which requests the entire program to be stopped. When `stop()` is called, any ongoing reactions at the current tick (t,n) are allowed to finish, and at the _next_ microstep the top-level reactor gets deleted. This means that all reactors see a `shutdown` event at (t, n+1).
